File: meshcall-ts/src/protocol.ts
import type { RawData } from "ws";
import { Ajv2020 } from "ajv/dist/2020.js";

export const PROTOCOL_VERSION = "meshcall/1" as const;

export type StreamKind =
  | "unary"
  | "server_stream"
  | "client_stream"
  | "duplex";

export type BalanceKind =
  | "round_robin"
  | "least_inflight"
  | "random"
  | "sticky"
  | "disabled";

export interface BalancePolicy {
  kind: BalanceKind;
  key?: string | null;
}

export interface HelloFrame {
  kind: "hello";
  protocol: string;
  role: "client" | "server";
  peer_id: string;
  instance_id?: string | null;
}

export interface HelloAckFrame {
  kind: "hello.ack";
  protocol: string;
  connection_id: string;
}

export interface RegisteredMethod {
  name: string;
  stream: StreamKind;
  balance: BalancePolicy;
}

export interface RegisteredService {
  name: string;
  methods: Array<RegisteredMethod>;
}

export interface ServerRegisterFrame {
  kind: "server.register";
  instance_id: string;
  services: Array<RegisteredService>;
}

export interface ServerRegisterAckFrame {
  kind: "server.register.ack";
  instance_id: string;
}

export interface CallOpenFrame {
  kind: "call.open";
  call_id: string;
  service: string;
  method: string;
  payload: unknown;
  deadline_unix_ms?: number | null;
}

export interface CallResultFrame {
  kind: "call.result";
  call_id: string;
  payload?: unknown;
}

export interface ErrorPayload {
  code: string;
  message: string;
  retryable?: boolean;
  details?: Record<string, unknown> | null;
}

export interface CallErrorFrame {
  kind: "call.error";
  call_id: string;
  error: ErrorPayload;
}

export interface CallCancelFrame {
  kind: "call.cancel";
  call_id: string;
  reason?: string;
}

export interface StreamItemFrame {
  kind: "stream.item";
  call_id: string;
  direction: "client" | "server";
  sequence: number;
  payload: unknown;
}

export interface StreamEndFrame {
  kind: "stream.end";
  call_id: string;
  direction: "client" | "server";
}

export interface StreamWindowFrame {
  kind: "stream.window";
  call_id: string;
  direction: "client" | "server";
  credit: number;
}

export interface PingFrame {
  kind: "ping";
  nonce: string;
}

export interface PongFrame {
  kind: "pong";
  nonce: string;
}

export type Frame =
  | HelloFrame
  | HelloAckFrame
  | ServerRegisterFrame
  | ServerRegisterAckFrame
  | CallOpenFrame
  | CallResultFrame
  | CallErrorFrame
  | CallCancelFrame
  | StreamItemFrame
  | StreamEndFrame
  | StreamWindowFrame
  | PingFrame
  | PongFrame;

export function decodeFrame(data: RawData | string): Frame {
  const text = typeof data === "string" ? data
    : Array.isArray(data) ? Buffer.concat(data).toString()
    : data instanceof ArrayBuffer ? Buffer.from(data).toString() : data.toString();
  const value: unknown = JSON.parse(text);
  if (!isRecord(value) || typeof value.kind !== "string") {
    throw new Error("Invalid MeshCall frame");
  }
  const validate = validators.get(value.kind);
  if (validate === undefined || !validate(value)) {
    throw new Error("Invalid MeshCall " + value.kind + " frame");
  }
  return value as unknown as Frame;
}

export function encodeFrame(frame: Frame): string {
  return JSON.stringify(frame);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

const string = { type: "string" };
const nullableString = { type: ["string", "null"] };
const direction = { enum: ["client", "server"] };
const integer = { type: "integer", minimum: 0, maximum: Number.MAX_SAFE_INTEGER };
const balance = {
  type: "object",
  properties: {
    kind: { enum: ["round_robin", "least_inflight", "random", "sticky", "disabled"] },
    key: nullableString,
  },
  required: ["kind"],
  additionalProperties: false,
};
const method = {
  type: "object",
  properties: {
    name: string, stream: { enum: ["unary", "server_stream", "client_stream", "duplex"] },
    balance,
  },
  required: ["name", "stream", "balance"],
  additionalProperties: false,
};
const service = {
  type: "object",
  properties: { name: string, methods: { type: "array", items: method } },
  required: ["name", "methods"],
  additionalProperties: false,
};
const ajv = new Ajv2020({ strict: false });
const frame = (kind: string, properties: Record<string, unknown>, required: string[]) =>
  ajv.compile({
    type: "object", properties: { kind: { const: kind }, ...properties },
    required: ["kind", ...required], additionalProperties: false,
  });
const validators = new Map([
  ["hello", frame("hello", {
    protocol: string, role: { enum: ["client", "server"] }, peer_id: string,
    instance_id: nullableString,
  }, ["protocol", "role", "peer_id"])],
  ["hello.ack", frame("hello.ack", { protocol: string, connection_id: string },
    ["protocol", "connection_id"])],
  ["server.register", frame("server.register", {
    instance_id: string, services: { type: "array", items: service },
  }, ["instance_id", "services"])],
  ["server.register.ack", frame("server.register.ack", { instance_id: string }, ["instance_id"])],
  ["call.open", frame("call.open", {
    call_id: string, service: string, method: string, payload: {},
    deadline_unix_ms: { type: ["integer", "null"] },
  }, ["call_id", "service", "method", "payload"])],
  ["call.result", frame("call.result", { call_id: string, payload: {} }, ["call_id"])],
  ["call.error", frame("call.error", {
    call_id: string,
    error: {
      type: "object",
      properties: {
        code: string, message: string, retryable: { type: "boolean" },
        details: { type: ["object", "null"] },
      },
      required: ["code", "message"], additionalProperties: false,
    },
  }, ["call_id", "error"])],
  ["call.cancel", frame("call.cancel", { call_id: string, reason: string }, ["call_id"])],
  ["stream.item", frame("stream.item", {
    call_id: string, direction, sequence: integer, payload: {},
  }, ["call_id", "direction", "sequence", "payload"])],
  ["stream.end", frame("stream.end", { call_id: string, direction }, ["call_id", "direction"])],
  ["stream.window", frame("stream.window", {
    call_id: string, direction, credit: { ...integer, minimum: 1 },
  }, ["call_id", "direction", "credit"])],
  ["ping", frame("ping", { nonce: string }, ["nonce"])],
  ["pong", frame("pong", { nonce: string }, ["nonce"])],
]);

File: meshcall-router/internal/router/protocol.go
package router

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

const Protocol = "meshcall/1"

type balance struct {
	Kind string  `json:"kind"`
	Key  *string `json:"key,omitempty"`
}

type method struct {
	Name    string  `json:"name"`
	Stream  string  `json:"stream"`
	Balance balance `json:"balance"`
}

type service struct {
	Name    string   `json:"name"`
	Methods []method `json:"methods"`
}

type errorPayload struct {
	Code      string          `json:"code"`
	Message   string          `json:"message"`
	Retryable bool            `json:"retryable"`
	Details   json.RawMessage `json:"details,omitempty"`
}

// Payload stays raw: the router never converts business numbers or field names.
type frame struct {
	Kind       string          `json:"kind"`
	CallID     string          `json:"call_id,omitempty"`
	Protocol   string          `json:"protocol,omitempty"`
	Role       string          `json:"role,omitempty"`
	PeerID     string          `json:"peer_id,omitempty"`
	Connection string          `json:"connection_id,omitempty"`
	InstanceID string          `json:"instance_id,omitempty"`
	Services   []service       `json:"services,omitempty"`
	Service    string          `json:"service,omitempty"`
	Method     string          `json:"method,omitempty"`
	Payload    json.RawMessage `json:"payload,omitempty"`
	Deadline   *int64          `json:"deadline_unix_ms,omitempty"`
	Error      *errorPayload   `json:"error,omitempty"`
	Reason     string          `json:"reason,omitempty"`
	Direction  string          `json:"direction,omitempty"`
	Sequence   *int64          `json:"sequence,omitempty"`
	Credit     *int64          `json:"credit,omitempty"`
	Nonce      string          `json:"nonce,omitempty"`
	raw        []byte
}

func decode(data []byte) (frame, error) {
	var f frame
	d := json.NewDecoder(bytes.NewReader(data))
	d.DisallowUnknownFields()
	if err := d.Decode(&f); err != nil {
		return f, errors.New("invalid frame")
	}
	if err := d.Decode(new(any)); err != io.EOF {
		return f, errors.New("expected one JSON frame")
	}
	switch f.Kind {
	case "hello":
		if f.Protocol != Protocol || (f.Role != "client" && f.Role != "server") || f.PeerID == "" {
			return f, errors.New("invalid hello or unsupported protocol")
		}
	case "server.register":
		if f.InstanceID == "" || f.Services == nil {
			return f, errors.New("invalid registration")
		}
	case "call.open":
		if f.Service == "" || f.Method == "" || f.Payload == nil {
			return f, errors.New("invalid call.open")
		}
	case "call.error":
		if f.Error == nil || f.Error.Code == "" {
			return f, errors.New("invalid call.error")
		}
	case "stream.item", "stream.end", "stream.window":
		if f.Direction != "client" && f.Direction != "server" {
			return f, errors.New("invalid stream direction")
		}
		if f.Kind == "stream.item" && (f.Sequence == nil || *f.Sequence < 0 || f.Payload == nil) {
			return f, errors.New("invalid stream item")
		}
		if f.Kind == "stream.window" && (f.Credit == nil || *f.Credit <= 0) {
			return f, errors.New("invalid stream credit")
		}
	case "call.result", "call.cancel", "ping", "pong":
	default:
		return f, fmt.Errorf("unexpected frame kind %q", f.Kind)
	}
	switch f.Kind {
	case "call.open", "call.result", "call.error", "call.cancel", "stream.item", "stream.end", "stream.window":
		if f.CallID == "" {
			return f, errors.New("missing call_id")
		}
	}
	f.raw = data
	return f, nil
}

func (f frame) bytes() []byte {
	if f.raw != nil {
		return f.raw
	}
	data, err := json.Marshal(f)
	if err != nil {
		panic(err) // Only internally constructed, JSON-safe control frames reach here.
	}
	return data
}

func callError(id, code, message string, retryable bool) frame {
	return frame{Kind: "call.error", CallID: id, Error: &errorPayload{Code: code, Message: message, Retryable: retryable}}
}

func (f frame) priority() bool {
	return f.Kind == "call.cancel" || f.Kind == "stream.window" || f.Kind == "ping" || f.Kind == "pong"
}
